ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -47,6 +48,359 @@ struct PendingRenameOperation
+    case MessageType_KtoU_NotifyFileHardLinkCreated:     return "NotifyFileHardLinkCreated";
+    case MessageType_KtoU_NotifyFilePreConvertToFull:    return "NotifyFilePreConvertToFull";
+    default:                                             return "Unknown";
+    };
+}
+
+class KextLogTracer
+{
+    bool discarded;
+    bool willEmitTrace;
+    bool hasTraceIndex;
+    char embeddedTraceBuffer[512];
+    char* dynamicTraceBuffer;
+    uint32_t dynamicTraceBufferSize;
+    uint32_t traceBufferPosition;
+    uint64_t traceIndex;
+
+public:
+    static RWLock traceFilterLock;
+    static _Atomic(char*) pathPrefixFilter;
+    static _Atomic(kauth_action_t) traceVnodeActionFilterMask;
+    static _Atomic(bool) traceDeniedVnodeEvents;
+    static _Atomic(bool) traceProviderMessagingEvents;
+    static _Atomic(bool) traceAllFileOpEvents;
+    static _Atomic(bool) traceAllVnodeEvents;
+    static _Atomic(uint64_t) nextTraceIndex;
+private:
+
+    KextLogTracer() = delete;
+    KextLogTracer(const NullTracer&) = delete;
+    KextLogTracer& operator=(const NullTracer&) = delete;
+
+    struct TraceBuffer
+    {
+        char* buffer;
+        uint32_t size, position;
+    };
+    
+    static bool ShouldTraceEventsForVnode(vnode_t vnode, char (&pathBuffer)[PATH_MAX])
+    {
+        bool shouldTrace = true;
+        if (KextLogTracer::pathPrefixFilter != nullptr)
+        {
+            int pathLen = sizeof(pathBuffer);
+            errno_t error = vn_getpath(vnode, pathBuffer, &pathLen);
+            if (error == 0)
+            {
+                RWLock_AcquireShared(KextLogTracer::traceFilterLock);
+                {
+                    if (KextLogTracer::pathPrefixFilter != nullptr)
+                    {
+                        shouldTrace = strprefix(pathBuffer, KextLogTracer::pathPrefixFilter);
+                    }
+                }
+                RWLock_ReleaseShared(KextLogTracer::traceFilterLock);
+            }
+            else
+            {
+                snprintf(pathBuffer, sizeof(pathBuffer), "[vn_getpath() failed: %d]", error);
+            }
+        }
+        
+        return shouldTrace;
+    }
+    
+    uint64_t GetTraceIndex()
+    {
+        if (!this->hasTraceIndex)
+        {
+            this->traceIndex = atomic_fetch_add(&KextLogTracer::nextTraceIndex, 1llu);
+            this->hasTraceIndex = true;
+        }
+        return this->traceIndex;
+    }
+    
+    TraceBuffer GetBuffer()
+    {
+        if (this->dynamicTraceBuffer != nullptr)
+        {
+            return TraceBuffer { this->dynamicTraceBuffer, this->dynamicTraceBufferSize, this->traceBufferPosition };
+        }
+        else
+        {
+            return TraceBuffer { this->embeddedTraceBuffer, sizeof(this->embeddedTraceBuffer), this->traceBufferPosition };
+        }
+    }
+    
+    bool GrowBuffer(uint32_t minimumSize)
+    {
+        TraceBuffer currentBuffer = this->GetBuffer();
```

Comment:
Do we have to worry about multiple threads here?

Reply:
No, each `KextLogTracer` instance has the lifetime of a vnode/fileop handler callback, and we stay on one thread for that.